src/lib/operators/delete.cpp

```diff
@@ -92,17 +88,19 @@ std::shared_ptr<const Table> Delete::_on_execute(std::shared_ptr<TransactionCont
 void Delete::_on_commit_records(const CommitID commit_id) {
   const auto chunk_count = _referencing_table->chunk_count();
   for (auto referencing_chunk_id = ChunkID{0}; referencing_chunk_id < chunk_count; ++referencing_chunk_id) {
-    const auto referencing_chunk = _referencing_table->get_chunk(referencing_chunk_id);
-    const auto referencing_segment =
-        std::static_pointer_cast<const ReferenceSegment>(referencing_chunk->get_segment(ColumnID{0}));
-    const auto referenced_table = referencing_segment->referenced_table();
+    const auto& referencing_chunk = _referencing_table->get_chunk(referencing_chunk_id);
+    const auto& referencing_segment =
+        static_cast<const ReferenceSegment&>(*referencing_chunk->get_segment(ColumnID{0}));
+    const auto referenced_table = referencing_segment.referenced_table();
 
-    for (const auto row_id : *referencing_segment->pos_list()) {
-      const auto referenced_chunk = referenced_table->get_chunk(row_id.chunk_id);
+    for (const auto row_id : *referencing_segment.pos_list()) {
+      const auto& referenced_chunk = referenced_table->get_chunk(row_id.chunk_id);
+      auto& mvcc_data = *referenced_chunk->mvcc_data();
 
-      referenced_chunk->mvcc_data()->set_end_cid(row_id.chunk_offset, commit_id);
-      referenced_chunk->increase_invalid_row_count(ChunkOffset{1});
+      mvcc_data.set_end_cid(row_id.chunk_offset, commit_id);
       // We do not unlock the rows so subsequent transactions properly fail when attempting to update these rows.
+      referenced_chunk->increase_invalid_row_count(ChunkOffset{1});
+      set_atomic_max(mvcc_data.max_end_cid, commit_id);
```

Comment:
Just realized we do this for each row. We should be able to do this after the loop, right? And then like:
```c++
referenced_chunk->increase_invalid_row_count(ChunkOffset{pos_list.size()});
set_atomic_max(mvcc_data.max_end_cid, commit_id);
```
@Bouncner thoughts on that?

Reply:
Hmm, please help me to understand the problem.

The "worst case" would be two concurrent delete operations, one deleting odd row IDs, the other even rowIDs, right?
Could it ever be a problem that a third concurrent operation reads an "not-yet-updated" `max_end_cid`? If not, I think we don't have to do it in the loop.

Reply:
Let's discuss this offline today together with these two other things:
- https://github.com/hyrise/hyrise/pull/2590#issuecomment-1648621053
- https://github.com/hyrise/hyrise/pull/2548#issuecomment-1640947399